linera-rpc/src/cross_chain_message_queue.rs

```diff
@@ -0,0 +1,199 @@
+};
+
+use futures::{channel::mpsc, StreamExt as _};
+use linera_base::identifiers::ChainId;
+use linera_core::data_types::CrossChainRequest;
+use rand::Rng as _;
+use tracing::{trace, warn};
+
+use crate::config::ShardId;
+
+#[cfg(with_metrics)]
+static CROSS_CHAIN_MESSAGE_TASKS: std::sync::LazyLock<prometheus::IntGauge> =
+    std::sync::LazyLock::new(|| {
+        prometheus::register_int_gauge!(
+            "cross_chain_message_tasks",
+            "Number of concurrent cross-chain message tasks",
+        )
+        .expect("IntGauge can be created")
+    });
+
+#[expect(clippy::too_many_arguments)]
+pub(crate) async fn forward_cross_chain_queries<F, G>(
+    nickname: String,
+    cross_chain_max_retries: u32,
+    cross_chain_retry_delay: Duration,
+    cross_chain_sender_delay: Duration,
+    cross_chain_sender_failure_rate: f32,
+    this_shard: ShardId,
+    mut receiver: mpsc::Receiver<(CrossChainRequest, ShardId)>,
+    handle_request: F,
+) where
+    F: Fn(ShardId, CrossChainRequest) -> G + Send + Clone + 'static,
+    G: Future<Output = anyhow::Result<()>>,
+{
+    let mut futures = futures::stream::FuturesUnordered::new();
+    let mut job_states: HashMap<QueueId, Job> = HashMap::new();
+
+    let run_task = |task: Task| async move { handle_request(task.shard_id, task.request).await };
+
+    let run_action = |action, queue, state: Job| async move {
+        linera_base::time::timer::sleep(cross_chain_sender_delay).await;
+
+        let to_shard = state.task.shard_id;
+
+        (
+            queue,
+            match action {
+                Action::Proceed { .. } => {
+                    if let Err(error) = run_task(state.task).await {
+                        warn!(
+                            nickname = state.nickname,
+                            %error,
+                            retry = state.retries,
+                            from_shard = this_shard,
+                            to_shard,
+                            "Failed to send cross-chain query",
+                        );
+
+                        Action::Retry
+                    } else {
+                        trace!(from_shard = this_shard, to_shard, "Sent cross-chain query",);
+
+                        Action::Proceed {
+                            id: state.id.wrapping_add(1),
+                        }
+                    }
+                }
+
+                Action::Retry => {
+                    linera_base::time::timer::sleep(cross_chain_retry_delay * state.retries).await;
+                    Action::Proceed { id: state.id }
+                }
+            },
+        )
+    };
+
+    loop {
+        tokio::select! {
+            Some((queue, action)) = futures.next() => {
+                let Entry::Occupied(mut state) = job_states.entry(queue) else {
+                    panic!("running job without state");
+                };
+
+                let remove = matches!(action, Action::Proceed { id } if state.get().id < id)
```

Comment:
TBH it wasn't easy - at least for me - to decipher why this particular `matches!` means we can remove the job from the queue. IIUC we remove it when this task's is the last one to arrive on this queue and we successfuly handled it. Maybe add a function on `state` (which is a `Job` 🙃 ) which would make it clear (even from the function name) why this queue was processed successfully.